code/model/modelread.cpp

```diff
@@ -5544,4 +5544,10 @@ void model_subsystem::reset()
 
     turret_max_bomb_ownage = 0;
     turret_max_target_ownage = 0;
+
+	beam_warmdown_program = actions::ProgramSet();
```

Comment:
I guess it will be necessary to add a hook like this wherever we want to add program functionality?  I'm just worried about `beam_warmup_program` and `beam_firing_program` and any number of other programs being added that would increase overhead.

Reply:
That is correct. Every single potential program entry would need a separate program set member.

What overhead are you referring to and how would you improve it? I do not see either thing being a problem since unused program hooks will just be an empty vector.

Reply:
I'm speaking very generally: overhead in execution time and overhead in memory usage.  The new hooks will require a certain amount of storage space and will need to be checked every frame.  Both of these will apply to every subsystem on every ship in the mission.

I don't actually know how consequential the overhead would be.  That being said, I have a hunch that nowadays a new MVP release will add incremental overhead that far exceeds the incremental overhead from this sort of thing.

Reply:
Memory overhead would only involve the actions parsed at startup which would be constant after that so there will be no significant allocations at runtime because of this system. If a program is triggered, new program locals are allocated for that but the memory required for that is pretty minimal. If no program is defined (as would be the case for every current mod) then only an empty vector would be allocated which takes up pretty much no memory.

For CPU overhead, if no program is defined then there is no additional overhead "per-frame" so existing mods would not cause any significant difference. For a running program, _once it is triggered_, we check every frame if things need to happen but that is only limited to the runtime of the program. If a program is triggered only as a response to an event happening then there is no overhead for that either.

Reply:
Well, that all sounds very good then.